server/utils/verifiers/VerifierBase.ts

```diff
@@ -0,0 +1,350 @@
+        if(subject.idIdentifierPreferred) {
+            const preferredIdentifier: DBAPI.Identifier | null = await DBAPI.Identifier.fetch(subject.idIdentifierPreferred);
+            if(!preferredIdentifier){
+                LOG.error(`subject's preferredId not found in the DB (id: ${subject.idSubject} | subject: ${subject.Name})`, LOG.LS.eTEST);
+                // subjectStats[i].isValid = false;
+            } else {
+                // grab our identifier details (type) and store it
+                const preferredIdentifierDetails: IdentifierDetails | null = await this.getIdentifierType(preferredIdentifier);
+                if(preferredIdentifierDetails) {
+                    result.preferred = preferredIdentifierDetails;
+                    result.details?.push(preferredIdentifierDetails);
+                }
+
+                // check for edan/ark and store as appropriate because the preferred id may be ARK too
+                switch(preferredIdentifierDetails?.identifierTypeEnum){
+                    case COMMON.eVocabularyID.eIdentifierIdentifierTypeEdanRecordID: {
+                        if(!result.edan) result.edan = preferredIdentifierDetails;
+                    } break;
+
+                    case COMMON.eVocabularyID.eIdentifierIdentifierTypeARK: {
+                        if(!result.ark) result.ark = preferredIdentifierDetails;
+                    } break;
+                }
+            }
+        }
+
+        // grab our list of identifiers from the SystemObject id
+        const identifiers: DBAPI.Identifier[] | null = await DBAPI.Identifier.fetchFromSystemObject(systemObject.idSystemObject);
+        if(!identifiers){
+            LOG.error(`could not get identifiers from subject (id: ${subject.idSubject} | subject: ${subject.Name})`, LOG.LS.eTEST);
+            return null;
+        }
+        if(identifiers.length<=0){
+            LOG.info(`(WARNING) no identifiers assigned to subject (id: ${subject.idSubject} | subject: ${subject.Name})`, LOG.LS.eTEST);
+            return result;
+        }
+
+        // cycle through all identifiers, find EDAN/ARK, and push to list
+        for(const identifier of identifiers) {
+
+            // get our details for this identifier, skip if error, store if valid
+            const details = await this.getIdentifierType(identifier);
+            if(!details) {
+                LOG.error(`could not get identifier details from subject (identifier: ${identifier.IdentifierValue} | id: ${subject.idSubject} | subject: ${subject.Name})`, LOG.LS.eTEST);
+                continue;
+            }
+
+            // make sure it doesn't already exist before pushing it
+            let idExists: boolean = false;
+            for(const id of result.details){
+                if(details.identifierTypeEnum===id.identifierTypeEnum &&
+                    details.identifier?.IdentifierValue==id.identifier?.IdentifierValue) {

```

Comment:
The use of ?. here on both sides of the equivalence test is interesting.  If both are undefined, than your test says you have a match.  Is that expected?

Reply:
If the types equal and the values are equal (even if they are undefined) then they are a match. Commented to revisit as there may be a better approach for the test.